File: src/log.cpp
#include "log.hpp"
#include "conf.hpp"
#include "exception.hpp"
#include <cerrno>
#include <chrono>
#include <cstring>
#include <iomanip>
#include <iostream>
#include <memory>
#include <sstream>

using namespace ftr;

void Log::init(const std::string prefix, std::shared_ptr<ftr::Conf> conf,
               bool log_stderr) {
    if (conf == nullptr) {
        throw ServerError("A conf object must be provided");
    }

    const std::string err_log_path = conf->get_error_log();

    if (err_log_path == "") {
        throw ServerError(
            "A path to the location of the error log must be provided");
    }

    const std::string acc_log_path = conf->get_access_log();

    if (acc_log_path == "") {
        throw ServerError(
            "A path to the location of the access log must be provided");
    }

    std::ios_base::openmode mode = std::ios_base::out;
    mode |= std::ios_base::app;

    // try opening the files
    m_err_log_stream.open(prefix + err_log_path, mode);

    if (!m_err_log_stream.is_open()) {
        throw ServerError(std::strerror(errno));
    }

    m_acc_log_stream.open(prefix + acc_log_path, mode);

    if (!m_acc_log_stream.is_open()) {
        throw ServerError(std::strerror(errno));
    }

    if (log_stderr) {
        m_log_stderr = true;
    }
}

Log::~Log() {
    if (m_acc_log_stream.is_open()) {
        m_acc_log_stream.close();
    }

    if (m_err_log_stream.is_open()) {
        m_err_log_stream.close();
    }
}

void Log::log_err(const std::string msg) {
    std::string cur_msg = log_msg(msg);

    m_err_log_stream.write(cur_msg.c_str(), cur_msg.size());
    m_err_log_stream.flush();

    if (m_log_stderr) {
        std::cerr << cur_msg;
    }
}

void Log::log_err(const std::string msg1, const std::string msg2) {
    log_err(msg1 + msg2);
}

void Log::log_acc(const std::string msg) {
    std::string cur_msg = log_msg(msg);

    m_acc_log_stream.write(cur_msg.c_str(), cur_msg.size());
    m_acc_log_stream.flush();

    if (m_log_stderr) {
        std::cerr << cur_msg;
    }
}

void Log::log_acc(const std::string msg1, const std::string msg2) {
    log_acc(msg1 + msg2);
}

std::string Log::log_msg(const std::string msg) {
    std::time_t cur_time =
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
    std::ostringstream ss;
    ss << '[';
    ss << std::put_time(std::localtime(&cur_time), "%Y/%m/%d %H:%M:%S");
    ss << "] ";
    ss << msg;
    ss << '\n';

    return ss.str();
}

File: src/log.rs
extern crate chrono;

use self::chrono::Local;
use ftr::Conf;
use std::fs::{File, OpenOptions};
use std::io::Write;
use std::sync::Mutex;

pub struct Log {
    log_stderr: bool,
    acc_log_handle: Mutex<File>,
    err_log_handle: Mutex<File>,
}

impl Log {
    pub fn new(prefix: &str, conf: &Conf, log_stderr: bool) -> std::io::Result<Self> {
        let prefix = String::from(prefix);
        let err_log_path = &conf.error_log;
        let acc_log_path = &conf.access_log;

        let err_file_handle = OpenOptions::new()
            .append(true)
            .create(true)
            .open(prefix.clone() + err_log_path)?;

        let acc_file_handle = OpenOptions::new()
            .append(true)
            .create(true)
            .open(prefix + acc_log_path)?;

        Ok(Log {
            log_stderr: log_stderr,
            acc_log_handle: Mutex::new(acc_file_handle),
            err_log_handle: Mutex::new(err_file_handle),
        })
    }

    pub fn log_err(&self, msg: &str) {
        let mut log = match self.err_log_handle.lock() {
            Ok(l) => l,
            Err(_err) => {
                return;
            }
        };
        let cur_msg = self.log_msg(msg);
        let _ = log.write(cur_msg.as_bytes());

        if self.log_stderr {
            eprint!("{}", cur_msg);
        }
    }

    pub fn log_acc(&self, msg: &str) {
        let mut log = match self.acc_log_handle.lock() {
            Ok(l) => l,
            Err(_err) => {
                return;
            }
        };
        let cur_msg = self.log_msg(msg);
        let _ = log.write(cur_msg.as_bytes());

        if self.log_stderr {
            eprint!("{}", cur_msg);
        }
    }

    fn log_msg(&self, msg: &str) -> String {
        let now = Local::now();
        let mut s = format!("[{}] ", now.format("%Y/%m/%d %H:%M:%S"));
        s.push_str(msg);
        s.push('\n');
        s
    }
}
